liquibase/liquibase
# Restore update testing rollback command

Fixes #1909 

Includes new integration test. I did some manual testing of it as well through the CLI.

Used the same arguments as `update` based on looking at old Main class

- - -
## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: [https://github.com/liquibase/liquibase/issues/1909](https://github.com/liquibase/liquibase/issues/1909)
* Local Branch: [https://github.com/liquibase/liquibase/tree/restore-update-testing-rollback](https://github.com/liquibase/liquibase/tree/restore-update-testing-rollback)
* Unit Tests: [https://github.com/liquibase/liquibase/pull/1939/checks](https://github.com/liquibase/liquibase/pull/1939/checks)
* Integration Tests: (shown in Unit Tests link above)
* Functional Tests: [https://jenkins.datical.net/job/liquibase-pro/job/restore-update-testing-rollback/](https://jenkins.datical.net/job/liquibase-pro/job/restore-update-testing-rollback/)

#### Testing
* Steps to Reproduce: [https://github.com/liquibase/liquibase/issues/1909](https://github.com/liquibase/liquibase/issues/1909)
* Guidance: 
  * (see above)

#### Dev Verification
(did not save output at the time)

#### Test Requirements (Internal Liquibase QA)
Although the reported issue is on Mac, this issue will exist on any OS.
The way updateTestingRollbackSQL works is:

```
Given a changelog with three change sets to deploy, the following liquibase operations are performed for each change set:
1. update changeset1; update changeset2; update changeset3
2. rollback changeset3; rollback changeset2; rollback changeset1
3. update changeset1; update changeset2 update changeset3
The thing to note is that if any of the changesets fail rollback, any changesets that come _after_ the failed changesets are not tested or updated. Any changesets that come _before _the failed changests are created on the database. 
```

_Manual_

* Verify a successful updateTestingRollback for a table that can be rolled back. :white_check_mark: 
  * At the end of the operation, the table is created on the database.
* Verify a failed updateTestingRollback for a table that cannot be rolled back. :white_check_mark:
  * The failure can be achieved by providing a custom rollback script that has a syntax error.
  * At the end of the operation the table is created on the database.
    * The object exists because the rollback itself fails.
* Verify a failed updateTestingRollback for three tables, table1, table2 and table3. table1 can be rolled back. table2 cannot be rolled back. table3 can be rolled back. :white_check_mark:
  * At the end of the operation, both table1 and table2 are created on the database.
  * At the end of the operation, table3 is not created on the database.
    * Processing of changes to deploy stops after the first rollback failure.
* Verify a successful updateTestingRollback for a stored logic object that can be rolled back. :white_check_mark:
  * Ensure a Pro license is provided.
  * At the end of the operation, the stored logic object is created on the database.
* Verify a failed updateTestingRollback for a stored logic object that cannot be rolled back and a table that can be rolled back. :white_check_mark:
  * The table is second in the list of changes to deploy.
  * At the end of the operation, the stored logic object is created on the database.
  * At the end of the operation table is not created on the database.

_Automated Functional Test_

* Add a new test class liquibase-pro-tests/[src](https://github.com/Datical/liquibase-pro-tests/tree/master/src)/[test](https://github.com/Datical/liquibase-pro-tests/tree/master/src/test)/[java](https://github.com/Datical/liquibase-pro-tests/tree/master/src/test/java)/[tests](https://github.com/Datical/liquibase-pro-tests/tree/master/src/test/java/tests)/[functional](https://github.com/Datical/liquibase-pro-tests/tree/master/src/test/java/tests/functional)/*update*/. Call the class updateTestingRollback.
* This does not need to be a multi-db platform test; do not use the Oracle platform as it is our slowest platform to test.
* Add a test to validate a successful updateTestingRollback.
* Add a test to validate a failed updateTestingRollback.



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-1910) by [Unito](https://www.unito.io)
